src/crab/ebpf_domain.hpp

```diff
@@ -209,33 +209,32 @@ class ebpf_domain_t final {
     bool get_map_fd_range(const Reg& map_fd_reg, int* start_fd, int* end_fd) const;
 
     struct TypeDomain {
-        void assign_type(NumAbsDomain& inv, const Reg& lhs, type_encoding_t t);
-        void assign_type(NumAbsDomain& inv, const Reg& lhs, const Reg& rhs);
-        void assign_type(NumAbsDomain& inv, const Reg& lhs, const std::optional<linear_expression_t>& rhs);
-        void assign_type(NumAbsDomain& inv, std::optional<variable_t> lhs, const Reg& rhs);
-        void assign_type(NumAbsDomain& inv, std::optional<variable_t> lhs, int rhs);
+        static void assign_type(NumAbsDomain& inv, const Reg& lhs, type_encoding_t t);
+        static void assign_type(NumAbsDomain& inv, const Reg& lhs, const Reg& rhs);
+        static void assign_type(NumAbsDomain& inv, const Reg& lhs, const std::optional<linear_expression_t>& rhs);
+        static void assign_type(NumAbsDomain& inv, std::optional<variable_t> lhs, const Reg& rhs);
+        static void assign_type(NumAbsDomain& inv, std::optional<variable_t> lhs, int rhs);
```

Comment:
Please don't. This is intentional. You may change the class to have a pointer to the numerical invariant as a member, if you prefer. 

The idea of these methods and the class is to make it easier to switch or extend the domain (for example, add a set domain) without affecting user code (or with less impact on user code).

Reply:
I can change it back with a bit of effort to make the functions callable from static functions in ebpf_domain_t but I'm not sure it makes it any easier to switch or extend the domain without affecting user code.

Reply:
The crux of the issue, as I understand it, is that `join_inv` is treated as part of ebpf_domain itself, even though it is essentially about the type (or about choosing variables based on types). Therefore it can only be called from `join_over_types` if it is static, which means `variable_has_type` must be static.

Is it not preferable then to rename `join_inv` to `selectively_join_based_on_type` or something along this line, and put it as a member inside `TypeDomain`?

Reply:
Ok, done